src/arizona_template_parser.erl

```diff
@@ -0,0 +1,287 @@
+-module(arizona_template_parser).
+-moduledoc false.
+
+%% --------------------------------------------------------------------
+%% API function exports
+%% --------------------------------------------------------------------
+
+-export([parse/1]).
+
+%
+
+-ignore_xref([parse/1]).
+
+%% --------------------------------------------------------------------
+%% Types (and their exports)
+%% --------------------------------------------------------------------
+
+-opaque element() :: {text, arizona_template_scanner:location(), binary()}
+                   | {expr, arizona_template_scanner:location(), binary()}
+                   | {block, arizona_template_scanner:location(), block()}
+                   | {tag, arizona_template_scanner:location(), tag()}.
+-export_type([element/0]).
+
+-type block() :: #{
+    name := binary(),
+    attributes := [attribute()]
+}.
+-export_type([block/0]).
+
+-type tag() :: #{
+    name := binary(),
+    is_stateful := boolean(),
+    is_void := boolean(),
+    attributes := [attribute()],
+    inner_content := [element()]
+}.
+-export_type([tag/0]).
+
+-type attribute_key() :: binary().
+-type attribute_value() :: {text, arizona_template_scanner:location(), binary()}
+                         | {expr, arizona_template_scanner:location(), binary()}.
+-type attribute() :: {attribute_key(), attribute_value()}.
```

Comment:
Is this purposefully not opaque?

Reply:
Yes, purposefully, so it is possible to pattern match against a map in the new compiler. However, it would be better to keep our code consistent by making it opaque. I'll change it.

Reply:
😄 We seem like two old guys with Alzheimer's.